### PR TITLE
[Quantization] Support compressed tensors w8a8 static and w8a8 dynamic weight

### PR DESCRIPTION
**Pull request by LHXuuu**
### What this PR does / why we need it?

While using the LLM Compressor quantization tool from the VLLM community to generate quantized weights, the VLLM Ascend engine needs to be adapted to support the compressed tensors quantization format.

1. Add AscendCompressedTensorsConfig to replace CompressedTensorsConfig in vllm.
2. Support CompressedTensorsW8A8 static weight.
   - weight: per-channel, int8, symmetric; activation: per-tensor, int8, symmetric.
4. Support CompressedTensorsW8A8Dynamic weight.
   - weight: per-channel, int8, symmetric; activation: per-token, int8, symmetric, dynamic.
5. Modify the override_quantization_method in AscendQuantConfig.

Co-authored-by: taoqun110 taoqun@huawei.com
Co-authored-by: chenxi-hh chen464822955@163.com

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

- vLLM version: v0.11.2
- vLLM main: https://github.com/vllm-project/vllm/commit/v0.11.2

### CODE REVIEW

**Comment by MengqingCao** [comment]
Thanks for this great work! Could you plz add an e2e test of w8a8 static and dynamic quant? And ut is also expected, but we could  add ut in the follow-up prs.

And is there any accuracy and performance mertics of your pr?

also cc @wangxiyuan @22dimensions

**Comment by MengqingCao** [comment]
You can solve the DCO and lint issues by referring to the contributing doc in https://vllm-ascend.readthedocs.io/

**Comment by LHXuuu** [comment]
> Thanks for this great work! Could you plz add an e2e test of w8a8 static and dynamic quant? And ut is also expected, but we could add ut in the follow-up prs.
> 
> And is there any accuracy and performance mertics of your pr?
> 
> also cc @wangxiyuan @22dimensions

Thanks for your reply. I’m currently running accuracy and performance tests. Once they’re complete, I’ll post them in the comment.

**Comment by LHXuuu** [comment]
@MengqingCao @wangxiyuan Hi! The precision results are shown in the table below. W8A8 static weights fall back to all down_proj linear, while w8a8 dynamic weights are fully quantized.
### Qwen3-32b precision test
| | ceval | gsm8k | mmlu |
|----|----|----|----|
| BF16 | 88.94 | 95.45 | 89.27 |
| w8a8 static | 89.03 | 95.45 | 88.91 |
| w8a8 dynamic | 88.72 | 96.51 | 89.16 |

File: vllm_ascend/quantization/compressed_tensors/compressed_tensors.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+QUANTIZATION_SCHEME_MAP_TYPE = dict[str, Optional[dict[str, QuantizationArgs]]]
+
+
+def remove_quantization_method():
+    if COMPRESSED_TENSORS_METHOD in QUANTIZATION_METHODS:
+        QUANTIZATION_METHODS.remove(COMPRESSED_TENSORS_METHOD)
+
+
+remove_quantization_method()
```
Comment (wangxiyuan) [inline vllm_ascend/quantization/compressed_tensors/compressed_tensors.py:38]: this is a hack way. We should let vLLM support quantization method overide  in the future.

File: vllm_ascend/worker/worker_v1.py [hunk 2]
```diff
@@ -157,6 +157,8 @@ def __init__(
         # FixMe: this is a patch to fix the issue cause by https://github.com/vllm-project/vllm/commit/de94289a98d7ec52a5ef02719e01a1db8b505170
         from vllm.model_executor.layers.linear import \
             WEIGHT_LOADER_V2_SUPPORTED
+        WEIGHT_LOADER_V2_SUPPORTED.append(
```
Comment (wangxiyuan) [inline vllm_ascend/worker/worker_v1.py]: ditto

File: vllm_ascend/quantization/compressed_tensors/schemes/compressed_tensors_w8a8.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
+#
+# Copyright (c) 2025 Huawei Technologies Co., Ltd. All Rights Reserved.
```
Comment (wangxiyuan) [inline vllm_ascend/quantization/compressed_tensors/schemes/compressed_tensors_w8a8.py]: still using huawei?

File: vllm_ascend/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_dynamic.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        # WEIGHT SCALE
+        weight_scale = ChannelQuantScaleParameter(
+            data=torch.empty((output_size_per_partition, 1),
+                             dtype=params_dtype),
+            output_dim=0,
+            weight_loader=weight_loader,
+        )
+        layer.register_parameter("weight_scale", weight_scale)
+
+    def apply_weights(self, layer: torch.nn.Module, x: torch.Tensor,
```
Comment (linfeng-yuan) [inline vllm_ascend/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_dynamic.py]: Could you please re-use `apply` and `process_weights_afeter_loading` functions of LinearMethod in `quantization/w8a8.py` and `quantization/w8a8_dynamic.py`. I think this can improve the maintainability, and ensure the accuracy and performance of quantized weights using modelslim and llm_compressor as much as possible.

File: vllm_ascend/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_dynamic.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    CompressedTensorsScheme
+from vllm.model_executor.parameter import (ChannelQuantScaleParameter,
+                                           ModelWeightParameter)
+
+from vllm_ascend.utils import ACL_FORMAT_FRACTAL_NZ, is_enable_nz
+
+logger = init_logger(__name__)
+
+
+class CompressedTensorsW8A8Dynamic(CompressedTensorsScheme):
```
Comment (NSDie) [inline vllm_ascend/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_dynamic.py]: Can it be normalized with the CompressedTensorsW8A8Dynamic class？

File: vllm_ascend/quantization/compressed_tensors/compressed_tensors.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+            weight_quant = scheme_dict.get("weights")
+            input_quant = scheme_dict.get("input_activations")
+
+        if weight_quant is None:
+            logger.warning_once("Acceleration for non-quantized schemes is "
+                                "not supported by Compressed Tensors. "
+                                "Falling back to UnquantizedLinearMethod")
+            return None
+
+        else:
```
Comment (Ronald1995) [inline vllm_ascend/quantization/compressed_tensors/compressed_tensors.py:199]: redundant else, you could remove it.

File: vllm_ascend/quantization/compressed_tensors/compressed_tensors.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+        # Map from [target -> scheme]
+        self.target_scheme_map = target_scheme_map
+        self.quant_description = config
+
+    def get_name(self) -> str:
+        return "compressed-tensors"
+
+    @classmethod
+    def get_supported_act_dtypes(cls) -> list[torch.dtype]:
+        return [torch.int8, torch.float16, torch.bfloat16]
```
Comment (Ronald1995) [inline vllm_ascend/quantization/compressed_tensors/compressed_tensors.py:63]: it's good to return `(torch.int8, torch.float16, torch.bfloat16)`

File: vllm_ascend/quantization/compressed_tensors/compressed_tensors.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+
+    @classmethod
+    def get_min_capability(cls) -> int:
+        raise NotImplementedError(
+            "Ascend hardware dose not support \"get_min_capability\" feature.")
+
+    @classmethod
+    def get_config_filenames(cls) -> list[str]:
+        return []
```
Comment (Ronald1995) [inline vllm_ascend/quantization/compressed_tensors/compressed_tensors.py:72]: it's good to return ()

File: vllm_ascend/quantization/compressed_tensors/compressed_tensors.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+        is_tensor = (weight_strategy and input_quant.strategy
+                     == QuantizationStrategy.TENSOR.value)
+        is_static = not weight_quant.dynamic and not input_quant.dynamic
+        is_symmetric = weight_quant.symmetric and input_quant.symmetric
+
+        # Only symmetric input quantization supported.
+        # Only symmetric weight quantization supported.
+        return is_8_bits and is_tensor and is_symmetric and is_static
+
+    def _is_dynamic_token_w8a8(self, weight_quant: QuantizationArgs,
```
Comment (Ronald1995) [inline vllm_ascend/quantization/compressed_tensors/compressed_tensors.py:235]: `_is_dynamic_token_w8a8` and `_is_static_tensor_w8a8` is duplicated, please extract a method

File: vllm_ascend/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_dynamic.py (hunk 4 above)
Comment (chenxi-hh) [reply vllm_ascend/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_dynamic.py]: ok,llm-compressor has been updated to reuse the modelslim code.

File: vllm_ascend/worker/worker_v1.py (hunk 2 above)
Comment (chenxi-hh) [reply vllm_ascend/worker/worker_v1.py]: removed

File: vllm_ascend/quantization/compressed_tensors/schemes/compressed_tensors_w8a8.py (hunk 3 above)
Comment (chenxi-hh) [reply vllm_ascend/quantization/compressed_tensors/schemes/compressed_tensors_w8a8.py]: removed, updated to reuse the modelslim code.

File: vllm_ascend/quantization/w8a8.py [hunk 10]
```diff
@@ -25,7 +25,9 @@
```
Comment (SlightwindSec) [inline vllm_ascend/quantization/w8a8.py:25]: Does it support ACLGraph?

Comment (chenxi-hh) [reply vllm_ascend/quantization/w8a8.py:25]: Yes, supported.

File: docs/source/user_guide/feature_guide/quantization-llm-compressor.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
+# Quantization Guide
```
Comment (MengqingCao) [inline docs/source/user_guide/feature_guide/quantization-llm-compressor.md]:
```suggestion
# LLM Compressor Quantization Guide
```

File: docs/source/user_guide/feature_guide/quantization-llm-compressor.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+To quantize a model, you should install [llm-compressor](https://github.com/vllm-project/llm-compressor/blob/main/README.md). It is a unified library for creating compressed models for faster inference with vLLM.
+
+Install llm-compressor:
+
+```bash
+pip install llmcompressor
+```
+
+### Generate the W8A8 weights
```
Comment (MengqingCao) [inline docs/source/user_guide/feature_guide/quantization-llm-compressor.md:25]: Let's declare the usage limits of this pr, e.g., int4 is not supported...

File: requirements.txt [hunk 13]
```diff
@@ -16,6 +16,7 @@ torchvision
 wheel
 pandas-stubs
 opencv-python-headless<=4.11.0.86 # Required to avoid numpy version conflict with vllm
+compressed_tensors>=0.11.0
```
Comment (MengqingCao) [inline requirements.txt:19]: I prefer to move this to `requirements-dev.txt`

File: pyproject.toml [hunk 14]
```diff
@@ -22,6 +22,7 @@ requires = [
     "quart",
     "numba",
     "opencv-python-headless<=4.11.0.86", # Required to avoid numpy version conflict with vllm
+    "compressed_tensors>=0.11.0"
```
Comment (MengqingCao) [inline pyproject.toml:26]: Let's remove it

File: tests/e2e/multicard/test_quantization.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+    example_prompts = [
+        "Hello, my name is",
+    ]
+    max_tokens = 5
+    with VllmRunner(snapshot_download("neuralmagic/Qwen2.5-3B-quantized.w8a8"),
+                    tensor_parallel_size=2,
+                    max_model_len=4096,
+                    gpu_memory_utilization=0.8,
+                    enforce_eager=False) as vllm_model:
+        vllm_model.generate_greedy(example_prompts, max_tokens)
```
Comment (MengqingCao) [inline tests/e2e/multicard/test_quantization.py]: Could we add a check between the golden results and the output here? Thus we could make sure the accuracy is not broke

File: vllm_ascend/quantization/compressed_tensors/compressed_tensors.py (hunk 1 above)
Comment (MengqingCao) [reply vllm_ascend/quantization/compressed_tensors/compressed_tensors.py:38]: +1, is there any progress on this?

File: docs/source/user_guide/feature_guide/quantization-llm-compressor.md (hunk 12 above)
Comment (MengqingCao) [reply docs/source/user_guide/feature_guide/quantization-llm-compressor.md:25]: You can refer to https://docs.vllm.ai/projects/ascend/en/latest/tutorials/DeepSeek-V3.2-Exp.html to construct the doc

File: vllm_ascend/quantization/compressed_tensors/compressed_tensors.py (hunk 1 above)
Comment (chenxi-hh) [reply vllm_ascend/quantization/compressed_tensors/compressed_tensors.py:38]: vLLM does not support rewriting quantization methods. modifications to the vLLM code will be required in the future.

File: requirements.txt (hunk 13 above)
Comment (chenxi-hh) [reply requirements.txt:19]: It is required at runtime.

File: pyproject.toml (hunk 14 above)
Comment (chenxi-hh) [reply pyproject.toml:26]: It is required at runtime.

File: tests/e2e/multicard/test_quantization.py (hunk 15 above)
Comment (chenxi-hh) [reply tests/e2e/multicard/test_quantization.py]: Accuracy check added

**Comment by LHXuuu** [comment]
@MengqingCao @wangxiyuan Hello, this pr is ready to merge.
